core/src/main/scala/cats/Foldable.scala

```diff
@@ -280,4 +304,23 @@ object Foldable {
       Eval.defer(if (it.hasNext) f(it.next, loop()) else lb)
     loop()
   }
+
+  private sealed abstract class Source[A] {
+    def uncons: Option[(A, Source[A])]
+  }
+
+  private object Source {
+    def empty[A]: Source[A] = new Source[A] {
+      def uncons = None
+    }
+
+    def cons[A](a: A, src: Eval[Source[A]]): Source[A] = new Source[A] {
+      def uncons = Some((a, src.value))
+    }
+
+    def fromFoldable[F[_], A](fa: F[A])(implicit F: Foldable[F]): Source[A] =
+      F.foldRight[A, Source[A]](fa, Now(Source.empty[A]))((a, evalSrc) =>
+        Now(cons(a, evalSrc))
+      ).value
+  }
```

Comment:
This is probably me prematurely optimizing, but I'd be tempted to do a couple things:

- Make `empty` reuse a singleton either by making `Source` covariant or by giving it a `def widen[AA >: A]: Source[AA] = this.asInstanceOf[Source[AA]]` method.
- Possibly store a value of `Now(Source.empty)` around for reuse by `fromFoldable`.